src/audio/kpb.c

```diff
@@ -119,11 +123,212 @@ static void kpb_free(struct comp_dev *dev)
+		trace_kpb_error("Failed to allocate space for "
+				"KPB bufefrs");
+		return -EINVAL;
+	}
+
+	cd->his_buf_hp.end_addr = cd->his_buf_hp.sta_addr +
+	(MAX_BUFFER_SIZE - LPSRAM_SIZE);
+
+	cd->his_buf_lp.end_addr = cd->his_buf_lp.sta_addr +
+	LPSRAM_SIZE;
+
+#elif KPB_NO_OF_HISTORY_BUFFERS == 1
+
+	cd->his_buf_hp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+					  MAX_BUFFER_SIZE);
+
+	if (cd->his_buf_hp.sta_addr) {
+		trace_kpb_error("Failed to allocate space for "
+				"KPB bufefrs");
+	return -EINVAL;
+	}
+
+	cd->his_buf_hp.end_addr = cd->his_buf_hp.sta_addr + MAX_BUFFER_SIZE;
+
+#else
+#error "Wrong number of key phrase buffers configured"
+#endif
+
+	/* TODO: zeroes both buffers */
+
+	/* Initialize clients data */
+	for (i = 0; i < MAX_NO_OF_CLIENTS; i++)
+		cd->clients[i].state = KPB_CLIENT_UNREGISTERED;
+
+	return ret;
+}
+
+/**
+ * kpb_copy() - copy real time input stream into sink buffer,
+ * and in the same time buffers that input for
+ * later use by some of clients.
+ *
+ * @arg1:  kpb component.
+ *
+ * Return: integer representing either 0 - success
+ * or -EINVAL - failure.
+ */
+static int kpb_copy(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *kpb = comp_get_drvdata(dev);
+	struct comp_buffer *source;
+	struct comp_buffer *sink;
+
+	trace_kpb("kpb_copy()");
+
+	/* Get source and sink buffers */
+	source = list_first_item(&dev->bsource_list, struct comp_buffer,
+				 source_list);
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       sink_list);
+	/* Process source data */
+	/* Check if source have data and valid pointer to read from */
+	if (source->avail > 0 && source && sink && source->r_ptr) {
+		/* Real time copying */
+		if (sink->free >= source->avail && sink->w_ptr) {
+			/* Sink and source are both ready and have space */
+			memcpy(sink->w_ptr, source->r_ptr,
+			       source->avail);
+
+			/* update source & snik data*/
+			comp_update_buffer_produce(sink, source->avail);
+			comp_update_buffer_consume(source, source->avail);
+		} else {
+			;/* What should we do if sink isn't ready(state/size)?
+			  * Simple skip is OK?
+			  */
+		}
+		/* Buffer source data internally for future use by clients */
+		if (source->avail <= MAX_BUFFER_SIZE)
+			kpb_buffer_data(kpb, source);
+	} else
+		ret = -EINVAL;
+
+	return ret;
+}
+
+static void kpb_buffer_data(struct comp_data *kpb, struct comp_buffer *source)
+{
+}
+
+/**
+ * kpb_cmd() - kpb event handler.
+ * @arg1:  kpb component.
+ * @arg2:  command to be executed
+ * @arg3:  pointer to data related with request
+ * @arg4:  maxiumum size of data
+ *
+ *
+ * Return: integer representing either 0 - success
+ * or -EINVAL - failure.
+ */
+static int kpb_cmd(struct comp_dev *dev, int cmd, void *data,
+		   int max_data_size)
+{
+	int ret = 0;
+	struct sof_ipc_ctrl_data *cdata = data;
+	struct kpb_client *cli = (struct kpb_client *)cdata->data->data;
+	uint8_t client_id = cli->id;
+
+	trace_kpb("kpb_cmd()");
+
+	switch (cmd) {
```

Comment:
@mrajwa is the expectation that these commands will be set from the driver? Generally these cmds are top level comp commands such as set_data/get_data etc. 
And what is the rationale for using sof_ipc_ctrl_data here? ctrl cmds are of type VOLUME/SWITCH/ENUM/BINARY typically

Reply:
First of all this is the place where I have most of my doubts. Somewhere in KPB module we need to handle requests specific for this module: add output sink (kpb_register_client) and start of draining (kpb_begin_draining) for example. These requests, needs to be triggered from outside of the KPB module and needs to be accompanied by DATA appropriate for this request. So for example *kpb_register_client* event needs client's/sink related data like draining history, state, id. So yes I though "int cmd" would contain request, and void *data its related DATA.
Now, can we simplify thing by agree "void *data" would be off type kpb_client instead of struct sof_ipc_ctrl_data?

Reply:
@keyonjie, @xiulipan, can you comment on this issue?